bundles/org.openhab.core.ui/src/main/java/org/openhab/core/ui/internal/components/UIComponentSitemapProvider.java

```diff
@@ -0,0 +1,306 @@
+                widget = defaultWidget;
+                setWidgetPropertyFromComponentConfig(widget, component, "height", SitemapPackage.DEFAULT__HEIGHT);
+                break;
+            default:
+                logger.warn("Unknown sitemap component type {}", component.getType());
+                break;
+        }
+
+        if (widget != null) {
+            setWidgetPropertyFromComponentConfig(widget, component, "label", SitemapPackage.WIDGET__LABEL);
+            setWidgetPropertyFromComponentConfig(widget, component, "icon", SitemapPackage.WIDGET__ICON);
+            setWidgetPropertyFromComponentConfig(widget, component, "item", SitemapPackage.WIDGET__ITEM);
+
+            if (widget instanceof LinkableWidget) {
+                LinkableWidget linkableWidget = (LinkableWidget) widget;
+                if (component.getSlots() != null && component.getSlots().containsKey("widgets")) {
+                    for (UIComponent childComponent : component.getSlot("widgets")) {
+                        Widget childWidget = buildWidget(childComponent);
+                        if (childWidget != null) {
+                            linkableWidget.getChildren().add(childWidget);
+                        }
+                    }
+                }
+            }
+
+            // TODO: process visibility & color rules
+        }
+
+        return widget;
+    }
+
+    private void setWidgetPropertyFromComponentConfig(Widget widget, UIComponent component, String configParamName,
+            int feature) {
+        if (component == null || component.getConfig() == null) {
+            return;
+        }
+        Object value = component.getConfig().get(configParamName);
+        if (value == null) {
+            return;
+        }
+        if (value instanceof Double) {
+            value = new BigDecimal((Double) value);
+        }
+        WidgetImpl widgetImpl = (WidgetImpl) widget;
+        widgetImpl.eSet(feature, value);
+    }
+
+    private void addWidgetMappings(EList<Mapping> mappings, UIComponent component) {
+        if (component.getConfig() != null && component.getConfig().containsKey("mappings")) {
+            if (component.getConfig().get("mappings") instanceof Collection<?>) {
+                for (Object sourceMapping : (Collection<?>) component.getConfig().get("mappings")) {
+                    if (sourceMapping instanceof String) {
+                        String cmd = sourceMapping.toString().split("=")[0].trim();
+                        String label = sourceMapping.toString().split("=")[1].trim();
+                        MappingImpl mapping = (MappingImpl) SitemapFactory.eINSTANCE.createMapping();
+                        mapping.setCmd(cmd);
+                        mapping.setLabel(label);
+                        mappings.add(mapping);
+                    }
+                }
+            }
+        }
+    }
+
+    @Override
+    public void addModelChangeListener(@NonNull ModelRepositoryChangeListener listener) {
```

Comment:
This class here should registered as a listener on `UIComponentRegistry`, then it could also inform any `ModelRepositoryChangeListener` that is passed in here about changes in the sitemap.